Admin/Admin.php

```diff
@@ -1,326 +1,368 @@
-        // replace admin and action name in pattern
-        $routeName = str_replace('{admin}', $this->underscore($this->getName()), $routingPattern);
-        $routeName = str_replace('{action}', $actionName, $routeName);
-
-        return $routeName;
+        // set current action
+        $action = $this->getAction($request->get('_route_params')['_action']);
+        $this->currentAction = $action;
+        // check if user is logged have required permissions to get current action
+        $this->checkPermissions($user);
+        // load entities according to action and request
+        $this->loadEntities($request);
     }
 
     /**
-     * @return string
-     */
-    public function getName()
-    {
-        return $this->name;
-    }
-
-    /**
-     * Return entity path for routing (for example, MyNamespace\EntityName => entityName)
+     * Check if user is allowed to be here
      *
-     * @return string
-     */
-    public function getEntityPath()
-    {
-        $array = explode('\\', $this->getEntityNamespace());
-        $path = array_pop($array);
-        $path = strtolower(substr($path, 0, 1)) . substr($path, 1);
-
-        return $path;
-    }
-
-    /**
-     * @return mixed
-     */
-    public function getEntityNamespace()
-    {
-        return $this->entityNamespace;
-    }
-
-    /**
-     * @return EntityRepository
-     */
-    public function getRepository()
-    {
-        return $this->repository;
-    }
-
-    /**
-     * @return mixed
-     */
-    public function getEntities()
-    {
-        return $this->entities;
-    }
-
-    /**
-     * @param mixed $entities
+     * @param UserInterface|string $user
      */
-    public function setEntities($entities)
+    public function checkPermissions($user)
     {
-        $this->entities = $entities;
+        if (!$user) {
+            throw new NotFoundHttpException('You must be logged to access to this url');
+        }
+        $roles = $user->getRoles();
+        $actionName = $this
+            ->getCurrentAction()
+            ->getName();
+
+        if (!$this->isActionGranted($actionName, $roles)) {
+            $message = sprintf('User with roles %s not allowed for action "%s"',
+                implode(', ', $roles),
+                $actionName
+            );
+            throw new NotFoundHttpException($message);
+        }
     }
 
     /**
-     * @return mixed
+     * Save entity via admin manager. Error are catch, logged and a flash message is added to session
+     *
+     * @return bool true if the entity was saved without errors
      */
-    public function getFormType()
+    public function save()
     {
-        return $this->formType;
+        try {
+            foreach ($this->entities as $entity) {
+                $this
+                    ->manager
+                    ->save($entity);
+            }
+            // inform user everything went fine
+            $this
+                ->messageHandler
```

Comment:
Could we have an accessor for this service and an interface for the MessageHandler?